packages/next/src/server/app-render/dynamic-rendering.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * The functions provided by this module are used to communicate certain properties
+ * about the currently running code so that Next.js can make decisions on how to handle
+ * the current execution in different rendering modes such as pre-rendering, resuming, and SSR.
+ *
+ * Today Next.js treats all code as potentially static. Certain APIs may only make sense when dynamically rendering.
+ * Traditionally this meant deopting the entire render to dynamic however with PPR we can now deopt parts
+ * of a React tree as dynamic while still keeping other parts static. There are really two different kinds of
+ * Dynamic indications.
+ *
+ * The first is simply an intention to be dynamic. unstable_noStore is an example of this where
+ * the currently executing code simply declares that the current scope is dynamic but if you use it
+ * inside unstable_cache it can still be cached. This type of indication can be removed if we ever
+ * make the default dynamic to begin with because the only way you would ever be static is inside
+ * a cache scope which this indication does not affect.
+ *
+ * The second is an indication that a dynamic data source was read. This is a stronger form of dynamic
+ * because it means that it is inappropriate to cache this at all. using a dynamic data source inside
+ * unstable_cache should error. If you want to use some dynamic data inside unstable_cache you should
+ * read that data outside the cache and pass it in as an argument to the cached function.
+ */
+
+// Once postpone is in stable we should switch to importing the postpone export directly
+import React from 'react'
+
+import type { StaticGenerationStore } from '../../client/components/static-generation-async-storage.external'
+import { DynamicServerError } from '../../client/components/hooks-server-context'
+
+import { StaticGenBailoutError } from '../../client/components/static-generation-bailout'
+
+const hasPostpone = typeof React.unstable_postpone === 'function'
+
+/**
+ * This function communicates that the current scope should be treated as dynamic.
+ *
+ * In most cases this function is a no-op but if called during
+ * a PPR prerender it will postpone the current sub-tree.
+ */
+export function markCurrentScopeAsDynamic(
```

Comment:
Nit: Curious if the term `scope` would be better replaced by something like `node`? Understanding that this _must_ be called within the _scope_ of a _component render_ as opposed to just called within app render would be helpful.